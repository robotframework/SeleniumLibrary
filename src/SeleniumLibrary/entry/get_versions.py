# Copyright 2008-2011 Nokia Networks
# Copyright 2011-2016 Ryan Tomac, Ed Manlove and contributors
# Copyright 2016-     Robot Framework Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from pathlib import Path
import re
import subprocess
import sys

from selenium import __version__

INSTALL_DIR = Path(__file__).parent.parent


def get_rf_version() -> str:
    process = subprocess.run(
        [sys.executable, "-m", "robot", "--version"], capture_output=True, check=False
    )
    return process.stdout.decode("utf-8").split(" ")[2]


def get_library_version() -> str:
    init_file = INSTALL_DIR / "__init__.py"
    with init_file.open("r") as file:
        data = file.read()
    return re.search('\n__version__ = "(.*)"', data).group(1)


def get_version():
    """Display Python, Robot Framework, SeleniumLibrary and selenium versions"""
    python_version = (
        f"{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}"
    )
    return (
        f"\nUsed Python is: {sys.executable}\n\tVersion: {python_version}\n"
        f'Robot Framework version: "{get_rf_version()}\n"'
        f"Installed SeleniumLibrary version is: {get_library_version()}\n"
        f"Installed selenium version is: {__version__}\n"
    )
