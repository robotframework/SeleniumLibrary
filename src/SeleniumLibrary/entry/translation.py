# Copyright 2008-2011 Nokia Networks
# Copyright 2011-2016 Ryan Tomac, Ed Manlove and contributors
# Copyright 2016-     Robot Framework Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import hashlib
import inspect
import json
from pathlib import Path
from typing import List, Optional

KEYWORD_NAME = "Keyword name"
DOC_CHANGED = "Documentation update needed"
NO_LIB_KEYWORD = "Keyword not found from library"
MISSING_TRANSLATION = "Keyword is missing translation"
MISSING_CHECKSUM = "Keyword tranlsaton is missing checksum"
MAX_REASON_LEN = max(
    len(DOC_CHANGED),
    len(NO_LIB_KEYWORD),
    len(MISSING_TRANSLATION),
    len(MISSING_CHECKSUM),
)


def get_library_translaton(plugings: Optional[str] = None) -> dict:
    from SeleniumLibrary import SeleniumLibrary

    selib = SeleniumLibrary(plugins=plugings)
    translation = {}
    for function in selib.attributes.values():
        translation[function.__name__] = {
            "name": function.__name__,
            "doc": function.__doc__,
            "sha256": hashlib.sha256(function.__doc__.encode("utf-16")).hexdigest(),
        }
    translation["__init__"] = {
        "name": "__init__",
        "doc": inspect.getdoc(selib),
        "sha256": hashlib.sha256(inspect.getdoc(selib).encode("utf-16")).hexdigest(),  # type: ignore
    }
    translation["__intro__"] = {
        "name": "__intro__",
        "doc": selib.__doc__,
        "sha256": hashlib.sha256(selib.__doc__.encode("utf-16")).hexdigest(),  # type: ignore
    }
    return translation


def _max_kw_name_lenght(project_tanslation: dict) -> int:
    max_lenght = 0
    for keyword_data in project_tanslation.values():
        if (current_kw_length := len(keyword_data["name"])) > max_lenght:
            max_lenght = current_kw_length
    return max_lenght


def _get_heading(max_kw_lenght: int) -> List[str]:
    heading = f"| {KEYWORD_NAME} "
    next_line = f"| {'-' * len(KEYWORD_NAME)}"
    if (padding := max_kw_lenght - len(KEYWORD_NAME)) > 0:
        heading = f"{heading}{' ' * padding}"
        next_line = f"{next_line}{'-' * padding}"
    reason = "Reason"
    reason_padding = MAX_REASON_LEN - len(reason)
    heading = f"{heading}| {reason}{' ' * reason_padding} |"
    next_line = f"{next_line} | {'-' * MAX_REASON_LEN} |"
    return [heading, next_line]


def _table_doc_updated(lib_kw: str, max_name_lenght: int, reason: str) -> str:
    line = f"| {lib_kw} "
    if (padding := max_name_lenght - len(lib_kw) - 0) > 0:
        line = f"{line}{' ' * padding}| {reason} "
    else:
        line = f"{line}| {reason} "
    if reason_padding := MAX_REASON_LEN - len(reason):
        line = f"{line}{' ' * reason_padding}"
    return f"{line}|"


def compare_translatoin(filename: Path, library_translation: dict):
    with filename.open("r") as file:
        project_translation = json.load(file)
    max_kw_lenght = _max_kw_name_lenght(library_translation)
    table_body = []
    for lib_kw, lib_kw_data in library_translation.items():
        project_kw_data = project_translation.get(lib_kw)
        if not project_kw_data:
            table_body.append(
                _table_doc_updated(lib_kw, max_kw_lenght, MISSING_TRANSLATION)
            )
            continue
        sha256_value = project_kw_data.get("sha256")
        if not sha256_value:
            table_body.append(
                _table_doc_updated(lib_kw, max_kw_lenght, MISSING_CHECKSUM)
            )
            continue
        if project_kw_data["sha256"] != lib_kw_data["sha256"]:
            table_body.append(_table_doc_updated(lib_kw, max_kw_lenght, DOC_CHANGED))
    for project_kw in project_translation:
        if project_kw not in library_translation:
            table_body.append(
                _table_doc_updated(project_kw, max_kw_lenght, NO_LIB_KEYWORD)
            )
    if not table_body:
        return []

    table = _get_heading(max_kw_lenght)
    table.extend(table_body)
    return table
