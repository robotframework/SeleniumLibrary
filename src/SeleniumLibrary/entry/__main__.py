# Copyright 2008-2011 Nokia Networks
# Copyright 2011-2016 Ryan Tomac, Ed Manlove and contributors
# Copyright 2016-     Robot Framework Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import json
from pathlib import Path
from typing import Optional
import click

from .get_versions import get_version
from .translation import compare_translatoin, get_library_translaton


CONTEXT_SETTINGS = {"help_option_names": ["-h", "--help"]}
VERSION = get_version()


@click.group()
@click.version_option(VERSION)
def cli():
    """Robot Framework SeleniumLibrary command line tool.

    Possible commands are:
    translation


    translation will generate detaul tranlsation json file from library keywords.

    See each command argument help for more details what (optional) arguments that command supports.
    """
    pass

@cli.command()
@click.argument(
    "filename",
    type=click.Path(exists=False, dir_okay=False, path_type=Path),
    required=True,
)
@click.option(
    "--plugings",
    help="Same as plugins argument in the library import.",
    default=None,
    type=str,
)
@click.option(
    "--compare",
    help="Compares the translation file sha256 sum to library documentation.",
    default=False,
    is_flag=True,
    show_default=True,
)
def translation(
    filename: Path,
    plugings: Optional[str] = None,
    compare: bool = False,
):
    """Default translation file from library keywords.

    This will help users to create their own translation as Python plugins. Command
    will populate json file with english language. To create proper translation
    file, users needs to translate the keyword name and doc arguments values in
    json file.

    The filename argument will tell where the default json file is saved.

    The --pluging argument is same as plugins argument in the library import.
    If you use plugins, it is also get default translation json file also witht
    the plugin keyword included in the library.

    If the --compare flag is set, then command does not generate template
    translation file. Then it compares sha256 sums from the filenane
    to ones read from the library documenentation. It will print out a list
    of keywords which documentation sha256 does not match. This will ease
    translation projects to identify keywords which documentation needs updating.
    """
    translation = get_library_translaton(plugings)
    if compare:
        if table := compare_translatoin(filename, translation):
            print(
                "Found differences between translation and library, see below for details."
            )
            for line in table:
                print(line)
        else:
            print("Translation is valid, no updated needed.")
    else:
        with filename.open("w") as file:
            json.dump(translation, file, indent=4)
        print(f"Translation file created in {filename.absolute()}")


if __name__ == "__main__":
    cli()
