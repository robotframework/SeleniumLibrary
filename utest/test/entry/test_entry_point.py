import json
from pathlib import Path
import sys

from approvaltests import verify_all

from SeleniumLibrary.entry.get_versions import get_version
from SeleniumLibrary.entry.translation import (
    compare_translatoin,
    get_library_translaton,
)


def test_version():
    lines = get_version().splitlines()
    python_version = (
        f"{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}"
    )
    assert len(lines) == 6
    assert "Used Python is: " in lines[1], lines[1]
    assert python_version in lines[2], lines[2]
    assert 'Robot Framework version: "' in lines[3], lines[3]
    assert "Installed SeleniumLibrary version is: " in lines[4], lines[4]
    assert "Installed selenium version is: " in lines[5], lines[5]


def test_get_translation():
    data = get_library_translaton()
    for item in data.values():
        assert item["name"], item["name"]
        assert item["doc"], item["doc"]
        assert item["sha256"], item["sha256"]


def test_compare_translation(tmp_path: Path):
    translation = tmp_path / "translation.json"
    data = get_library_translaton()
    with translation.open("w") as file:
        json.dump(data, file, indent=4)
    table = compare_translatoin(translation, data)
    verify_all("No changes", table)


def test_compare_translation_changes(tmp_path: Path):
    translation = tmp_path / "translation.json"
    data = get_library_translaton()
    data.pop("handle_alert", None)
    data["alert_should_be_present"]["sha256"] = "foo"
    with translation.open("w") as file:
        json.dump(data, file, indent=4)
    table = compare_translatoin(translation, get_library_translaton())
    verify_all("Changes", table)
