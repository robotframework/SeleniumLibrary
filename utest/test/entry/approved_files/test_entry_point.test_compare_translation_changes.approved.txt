Changes

0) | Keyword name                             | Reason                                 |
1) | ---------------------------------------- | -------------------------------------- |
2) | alert_should_be_present                  | Documentation update needed            |
3) | handle_alert                             | Keyword is missing translation         |
